File: solutions/python3/0430.py
class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        curr = head

        while curr:
            if curr.child:
                next = curr.next
                curr.next = curr.child
                curr.child.prev = curr
                curr.child = None
                tail = curr.next
                while tail.next:
                    tail = tail.next
                tail.next = next
                if next:
                    next.prev = tail
            curr = curr.next

        return head

File: solutions/cpp/0430.cpp
class Solution {
 public:
  Node* flatten(Node* head) {
    for (Node* curr = head; curr; curr = curr->next)
      if (curr->child) {
        Node* next = curr->next;
        curr->next = curr->child;
        curr->child->prev = curr;
        curr->child = nullptr;
        Node* tail = curr->next;
        while (tail->next) tail = tail->next;
        tail->next = next;
        if (next) next->prev = tail;
      }

    return head;
  }
};